docs/design-documents/features/storage/TDBStore/TDBStore_design.md

```diff
@@ -0,0 +1,497 @@
+
+Header:  
+`virtual int set_finalize(set_handle_t handle);`
+
+Pseudo code:
+- Advance `_free_space_offset` to padded offset 
+- Update a `record_header_t` structure with all relevant values
+- Program header at `bd_base_offset` from handle with pads
+- Call `sync` on buffered block device
+- Free `handle`
+- Release `_mutex`
+
+### Key iterator APIs
+
+**iterator_open function**
+
+Header:  
+`virtual int iterator_open(iterator_t *it, const char *prefix = NULL);`
+
+Pseudo code:
+- Take `_mutex`  
+- Allocate a `key_iterator_handle_t` structure into `it`
+- Set `ram_table_ind` field in iterator to 0
+- Duplicate `prefix` into same field in iterator
+- Release `_mutex`
+
+**iterator_next function**
+
+Header:  
+`virtual int iterator_next(iterator_t it, char *key, size_t key_size);`
+
+Pseudo code:
+- Take `_mutex`  
+- While `ram_table_ind` field in iterator smaller than `_num_keys`
+	- Read key pointed to by ram table in `ram_table_ind` into a local variable
+	- If name matches prefix
+		- Advance `ram_table_ind` field in iterator
+		- Copy name to `key` and return OK
+	- Advance `ram_table_ind` field in iterator 
+- Return "not found" error
+- Release `_mutex`
+
+**iterator_close function**
+
+Header:  
+`virtual int iterator_close(iterator_t it);`
+
+Pseudo code:
+- Release `prefix` field in iterator and structure allocated at `it`
+
+
+### Reserved space
+
+**reserved_space_set function**
+
+Header:  
+`virtual int reserved_space_set(void *data);`
+
+Pseudo code:  
+- Check if reserved space is not empty, if it is, return a "reserved space programmed error"
+- Copy `data` contents to reserved space location 
+
+**reserved_space_get function**
+
+Header:  
+`virtual int reserved_space_get(void *data);`
+
+Pseudo code:  
+- Copy contents from reserved space location `data`  
+
+
+# Usage scenarios and examples
+
+### Standard usage of the class
+
+Following example code shows standard usage of the TDBStore class 
+
+**Standard usage example**
+
+```C++
+// Underlying block device. Here, SPI Flash is fully used.
+// One can use SlicingBlockDevice if we want a partition.
+SPIFBlockDevice bd(PTE2, PTE4, PTE1, PTE5);
+
+// Instantiate tdbstore with our block device and 64 as max number of keys
+TDBStore tdbstore(64, &bd);
+
+int res;
+
+// Initialize tdbstore
+res = tdbstore.init();
+
+// Add "Key1"
+const char *val1 = "Value of key 1";
+const char *val2 = "Updated value of key 1";
+res = tdbstore.set("Key1", val1, sizeof(val1), 0);
+// Update value of "Key1"
+res = tdbstore.set("Key1", val2, sizeof(val2), 0);
+
+uint_8 value[32];
+size_t actual_size;
+// Get value of "Key1". Value should return the updated value.
+res = tdbstore.get("Key1", value, sizeof(value), &actual_size);
+
+// Remove "Key1"
+res = tdbstore.remove("Key1");
+
+// Incremental write, if need to generate large data with a small buffer
+const int data_size = 1024;
+char buf[8];
+
+KVSTore::set_handle_t handle;
+res = tdbstore.set_start(&handle, "Key2", data_size, 0);
```

Comment:
When this lands on the handbook, we may want an additional example that just shows off "get/set/remove". That would probably be the best way to show the simplicity of KVStore API and would make the best 1st impressions.

You're probably alread thinking that, but just wanted to mention it. We should have both examples since incremental set is also useful.